src/codegen/statements/try_catch.rs

```diff
@@ -439,47 +433,93 @@ fn insert_catch_clauses(
         left: ret_data_len.into(),
         right: selector_data_len.into(),
     };
-    cfg.add(
-        vartab,
-        Instr::BranchCond {
-            cond: cond_enough_data,
-            true_block: enough_data_block,
-            false_block: no_match_err_id,
-        },
-    );
-
-    let offset = Expression::NumberLiteral {
-        loc: Codegen,
-        ty: Uint(32),
-        value: 0.into(),
-    };
-    let err_selector = Expression::Builtin {
-        loc: Codegen,
-        tys: vec![Type::Bytes(4)],
-        kind: Builtin::ReadFromBuffer,
-        args: vec![buffer.clone(), offset],
-    };
-
-    cfg.set_basic_block(enough_data_block);
-    // Expect the returned data to match the 4 bytes function selector for "Error(string)"
-    let err_id = Expression::NumberLiteral {
-        loc: Codegen,
-        ty: Type::Bytes(4),
-        value: ERROR_SELECTOR.into(),
-    };
-    let cond = Expression::Equal {
-        loc: Codegen,
-        left: err_selector.into(),
-        right: err_id.into(),
-    };
-    let match_err_id = cfg.new_basic_block("match_err_id".into());
     let instruction = Instr::BranchCond {
-        cond,
-        true_block: match_err_id,
+        cond: cond_enough_data,
+        true_block: next_clause.unwrap(),
         false_block: no_match_err_id,
     };
     cfg.add(vartab, instruction);
 
+    for (n, clause) in try_stmt.errors.iter().enumerate() {
+        cfg.set_basic_block(next_clause.unwrap());
+        let next_index = n + 1;
+        next_clause = try_stmt
+            .errors
+            .get(next_index)
+            .map(|_| cfg.new_basic_block(format!("catch_error_{}", next_index)));
+
+        // Expect the returned data to match a known 4 bytes function selector
+        let err_id = Expression::NumberLiteral {
+            loc: Codegen,
+            ty: Type::Bytes(4),
+            value: match clause.param.as_ref().unwrap().ty {
+                Type::String => ERROR_SELECTOR.into(),
+                Type::Uint(256) => PANIC_SELECTOR.into(),
+                _ => unreachable!("Only 'Error(string)' and 'Panic(uint256)' can be caught"),
+            },
+        };
+        let offset = Expression::NumberLiteral {
+            loc: Codegen,
+            ty: Uint(32),
+            value: 0.into(),
+        };
+        let err_selector = Expression::Builtin {
+            loc: Codegen,
+            tys: vec![Type::Bytes(4)],
+            kind: Builtin::ReadFromBuffer,
+            args: vec![buffer.clone(), offset],
+        };
+        let cond = Expression::Equal {
+            loc: Codegen,
+            left: err_selector.into(),
+            right: err_id.into(),
+        };
+        let match_err_id = cfg.new_basic_block("match_err_id".into());
+        let instruction = Instr::BranchCond {
```

Comment:
This would a lot more efficient with `Instr::Switch`.

For every error case you're reading the value from the buffer, rather than keeping the value in a `Type::Uint(32)` or `Type::Bytes(4)` and switching on that.

Reply:
There are a maximum of two possible switch cases. So I don't think it's a lot more efficient even in the worst case.

This was done with a loop before, so I just adapted it. However I agree using a switch is cleaner, I can change that.

> For every error case you're reading the value from the buffer

Good catch, this is indeed unnecessary